docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1844,16 +1855,94 @@ entry value. When providing serializable enum values through `P4Data`, one can
 either use the enum entry's name (`enum` human-readable string field) or its
 assigned value (`enum_value` bytestring field).
 
+### User-defined types { #sec-user-defined-types}
+
+P4~16~ enables programmers to introduce new types [@P4NewTypes]. While similar
+to `typedef`, this mechanism introduces in fact a new type, which is not
+strictly synonym with the original type. It is important to preserve this
+distinction in the P4Info message, in particular for the purposes of
+[translation](#sec-psa-metadata-translation). When introducing a new type, the
+declaration can be annotated with `@p4runtime_translation` to indicate that the
+type exposed to the P4Runtime client is different from the original P4 type. One
+important use-case if for [port numbers](#sec-translation-of-port-numbers),
+whose underlying dataplane representation may vary on different targets, but for
+which it may be convenient to present a unified representation and numbering
+scheme to the control-plane. **The `@p4runtime_translation` annotation can only
+be used if the underlying P4 built-in type is a fixed-width unsigned bitstring
+type (`bit<W>`)** and the type exposed to the control-plane will also be a
+fixed-width unsigned bitstring, with a potentially different bitwidth. It takes
```

Comment:
I guess the restriction to `bit<W>` types might be generalized in later P4Runtime API spec versions after 1.0?

Reply:
Nothing off the top of my head.  I was more asking the question aloud whether we are thinking we are restricting ourselves from allowing it in the future.  I am pretty sure we are not restricting ourselves in that way, if/when someone comes up with a strong enough use case.

I do think that it is more likely that a P4 `type` might be used in the future with the underlying type being a struct, with no @p4runtime_translation annotation, e.g. for being able to name sub-fields of a larger value.  Again, as long as the P4 Info can be generalized so that original_type elsewhere in this proposal can be any P4 type we want to be able to handle, no problem.  Also not urgent for P4Runtime version 1.0.